api/src/main/java/org/wildfly/transformer/TransformerBuilder.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.transformer;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import static java.lang.Thread.currentThread;
+
+/**
+ * Resource transformer builder. Instances of this class are thread safe.
+ *
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opálka</a>
+ */
+public abstract class TransformerBuilder {
+    private static final String DEFAULT_CONFIG = "default.mapping";
+    private static final int MAX_MAPPINGS = 0xFFFF;
+    private static final char DOT = '.';
+    private static final char SEP = '/';
+    private final Thread thread;
+    private final Map<String, String> mappingWithSeps;
+    private final Map<String, String> mappingWithDots;
+    private InputStream mappingFile;
+    private boolean built;
+
+    protected TransformerBuilder() {
+        thread = currentThread();
+        mappingWithSeps = new HashMap<>();
+        mappingWithDots = new HashMap<>();
+    }
+
+    /**
+     * Adds custom packages mapping defined in a configuration file.
+     * First this method tries to read the specified file from the file system.
+     * If file doesn't exist on the file system then class loader resources are inspected.
+     * The specified file <code>config</code> must exist either on the  file system or
+     * in class loader resources otherwise exception is thrown.
+     * Once this method is called it turns off <i>default packages mapping configuration</i>
+     * and <i>user provided configuration</i> is used instead.
+     *
+     * @param config packages mapping configuration file
+     * @return this builder instance
+     * @throws ConcurrentModificationException if this builder instance is used by multiple threads
+     * @throws IllegalStateException if either this or {@link #build()} method have been already called
+     * @throws IllegalArgumentException if method parameter is <code>null</code>
+     * or if method parameter equals to <code>empty string</code>
+     * or if file doesn't exist neither on the file system nor in class loader resources
+     * @throws IOException if some I/O error occurs while reading the configuration file <code>config</code>
+     */
+    public final TransformerBuilder setPackagesMapping(final String config) throws IOException {
+        // preconditions
+        if (thread != currentThread()) throw new ConcurrentModificationException();
```

Comment:
It might be clearer if we include a message indicating why the exception is thrown, so that the source code doesn't have to be checked to understand why it might of been thrown.

Reply:
Fixed